hyperspy/_signals/signal2d.py

```diff
@@ -605,3 +608,433 @@ def add_ramp(self, ramp_x, ramp_y, offset=0):
+        return max_position
+
+    def distance(x, y):
+        """Calculates the distance between two points."""
+        v = x - y
+        return np.sqrt(np.sum(np.square(v)))
+
+    def gradient(image):
+        """Calculates the square of the 2-d partial gradient.
+
+        Parameters
+        ----------
+        image : ndarray
+
+        Returns
+        -------
+        ndarray
+
+        """
+        gradient_of_image = np.gradient(image)
+        gradient_of_image = gradient_of_image[0] ** 2 + gradient_of_image[
+                                                            1] ** 2
+        return gradient_of_image
+
+    # Generate an ordered list of matrix coordinates.
+    if len(z.shape) != 2:
+        raise ValueError("'z' should be a 2-d image matrix.")
+    z = z / np.max(z)
+    coordinates = np.indices(z.data.shape).reshape(2, -1).T
+    # Calculate the gradient at every point.
+    image_gradient = gradient(z)
+    # Boolean matrix of high-gradient points.
+    gradient_is_above_threshold = image_gradient >= grad_threshold
+    peaks = []
+    for coordinate in coordinates[gradient_is_above_threshold.flatten()]:
+        # Iterate over coordinates where the gradient is high enough.
+        b = box(coordinate[0], coordinate[1], window_size, z.shape[0],
+                z.shape[1])
+        p_old = np.array([0, 0])
+        p_new = get_max(z, b)
+        while np.all(p_old != p_new):
+            p_old = p_new
+            b = box(p_old[0], p_old[1], window_size, z.shape[0], z.shape[1])
+            p_new = get_max(z, b)
+        if distance(coordinate, p_new) <= distance_cutoff:
+            peaks.append(tuple(p_new))
+    peaks = np.array([np.array(p) for p in set(peaks)])
+    return clean_peaks(peaks)
+
+
+def find_peaks_stat(z, standard_deviation=1.):
+    """
+    Method to locate positive peaks in an image based on statistical refinement
+    and difference with respect to mean intensity.
+    Parameters
+    ----------
+    z : ndarray
+        Array of image intensities.
+    Returns
+    -------
+    ndarray
+        (n_peaks, 2)
+        Array of peak coordinates.
+    Notes
+    -----
+    Implemented as described in the PhD thesis of Thomas White (2009) the
+    algorithm was developed by Gordon Ball during a summer project in
+    Cambridge.
+    This version by Ben Martineau (2016), with minor modifications to the
+    original where methods were ambiguous or unclear.
+    """
+    from scipy.ndimage.filters import generic_filter
+    from scipy.ndimage.filters import uniform_filter
+    from sklearn.cluster import DBSCAN
+
+    def normalize(image):
+        """Scales the image to intensities between 0 and 1."""
+        return image / np.max(image)
+
+    def _local_stat(image, radius, func):
+        """Calculates rolling method 'func' over a circular kernel."""
+        x, y = np.ogrid[-radius:radius + 1, -radius:radius + 1]
+        kernel = x ** 2 + y ** 2 <= radius ** 2
```

Comment:
Could use `np.hypot` here - not sure if better (I would generally assume they do something smart, having included this)